### ISSUE TITLE
WASI support?

### ISSUE THREAD

**Issue by Restioson** [smol-rs/polling#102]
[Mio](https://github.com/tokio-rs/mio) has limited support for the `wasm32-wasi` target and I think it should therefore be possible to implement support here too, using WASI's `poll_oneoff` interface. This could be useful to allow for async TcpStreams and similar on wasm32-wasi rust binaries using `async-io`

**Comment by notgull** [comment]
There's [this branch](https://github.com/smol-rs/polling/tree/wasi/), but I'm not sure if @yoshuawuyts is actively working on it.

We could probably support WASI with the same caveats the `mio` has (i.e. it can't be woken since WASI is single threaded... although now that [WASI threads](https://bytecodealliance.org/articles/wasi-threads) are a thing, shouldn't there be a way around that?). The actual issue would show up down the line is in `async-io`, since we can't spawn the "`async-io`" thread yet. We would need to conditionally use a single threaded polling strategy there.

**Comment by notgull** [comment]
Sounds good, thanks for clarifying! It's probably better to wait for `preview2` to drop as opposed to trying to hammer `preview1` into a usable state.

**Comment by notgull** [comment]
`wasm32-wasi-preview2` has now [landed](https://doc.rust-lang.org/nightly/rustc/platform-support/wasm32-wasi-preview2.html) on Rust Nightly. So now it might be worth looking into this further.

**Comment by taiki-e** [comment]
> wasm32-wasi-preview2 has now [landed](https://doc.rust-lang.org/nightly/rustc/platform-support/wasm32-wasi-preview2.html) on Rust Nightly

(It is likely to be reverted once, though: https://github.com/rust-lang/rust/pull/120434)

**Comment by yoshuawuyts** [comment]
A new version of the `wasi` crate has been published containing [`poll`](https://docs.rs/wasi/latest/wasi/io/poll/fn.poll.html) operations. An example usage of these bindings can be found [here](https://github.com/lann/wasi-hyperium/blob/7eaeae4cc161ff097806ff0fb24207e4b1642469/src/poll.rs). It is possible to author and test this code already via [cargo-component](https://github.com/bytecodealliance/cargo-component), even if we don't yet have a Rustc target for WASI 0.2. We could for example put it behind a feature flag. Though the WASI 0.2 target is expected to be merged soon, and maybe by the time we've implemented this we're okay.

**Comment by notgull** [comment]
@yoshuawuyts Great! The only thing missing is a way to wake up the `Poller` from another thread. Although as WASI has no threads I don't think it's possible to do that yet.

**Comment by yoshuawuyts** [comment]
Finished writing a wasi native async runtime, and wrote down step by step instructions for how this all works here: https://blog.yoshuawuyts.com/building-an-async-runtime-for-wasi/.

If someone wants to integrate this into `polling`, they should be able to follow most of that and it should all work!

**Comment by vkrivopalov** [comment]
I started looking into adding WASI support to `polling`, and I am facing a conundrum here.

`WASIp1` supports polling through [`wasi::poll_oneoff`](https://docs.rs/wasi/0.11.0+wasi-snapshot-preview1/wasi/fn.poll_oneoff.html) using [`wasi::Subscription`](https://docs.rs/wasi/0.11.0+wasi-snapshot-preview1/wasi/struct.Subscription.html)s that can be created from raw file descriptors.
This is exactly how WASI support is implemented in [mio](https://github.com/tokio-rs/mio).

WASIp1 is supported by [wasi-rs](https://github.com/bytecodealliance/wasi-rs/) v0.11, versions 0.12 and up have completely different API and support WASIp2. It no longer uses `wasi::poll_oneoff` and instead relies on the [`Pollable`](https://docs.rs/wasi/0.13.3+wasi-0.2.2/wasi/io/poll/struct.Pollable.html) type as explained in @yoshuawuyts's awesome article linked above.

From the comments in this thread I conclude that it is preferred to implement support WASIp2 rather than WASIp1. This is possible to do even though there is some mismatch between the public `Poller` API and how `wasi::io::poll::Pollable` combines the source and the interest in certain types of events.
But as far as I can tell only WASI-specific types such as [`wasi::sockets::tcp::TcpSocket`](https://docs.rs/wasi/0.13.3+wasi-0.2.2/wasi/sockets/tcp/struct.TcpSocket.html) expose methods to obtain `Pollable`, it cannot be constructed for standard library types such as `std::net::TcpStream` or `std::net::TcpListener`.

This sort of defeats the purpose of using `polling` for WASI because `polling` provides an abstraction over specific platform and a portable interface.

What is best to be done then, should we try to implement support for WASIp1 which can be used for standard library types or integrate the `Pollable`-based polling from WASIp2 but limit its application to WASI-specific types? Please advise.

**Comment by notgull** [comment]
Yeah, that's a good point. There's not really a way to traditionally "register" a file descriptor into a `Poller` at this point. So, the `polling` model doesn't fit WASI.

@smol-rs/admins I'm planning on closing this issue as "not planned", since it's not likely that we'll be able to implement this.

**Comment by yoshuawuyts** [comment]
@dicej has been [doing work](https://github.com/tokio-rs/mio/pull/1836#issuecomment-2474524724) to upstream WASI 0.2 support to `mio`. From the sound of it this might result in changes to wasi-libc to provide backwards compatibility with the POSIX model. I believe this could be used by `polling` as well.

If I understand the plan correctly, this same shim would be guaranteed to keep working for WASI 0.3 as well, which should give this some longevity.

**Comment by dicej** [comment]
Correct.  We can lean on `wasi-libc` to make WASI look like just another POSIX OS as a good first default, then add WASI-specific improvements later as desired.
